Storage/src/Lifecycle.php

```diff
@@ -0,0 +1,236 @@
+<?php
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+/**
+ * Object Lifecycle Management supports common use cases like setting a Time to
+ * Live (TTL) for objects, archiving older versions of objects, or "downgrading"
+ * storage classes of objects to help manage costs.
+ *
+ * This builder does not execute any network requests and is intended to be used
+ * in combination with either
+ * {@see Google\Cloud\Storage\StorageClient::createBucket()}
+ * or {@see Google\Cloud\Storage\Bucket::update()}.
+ *
+ * Example:
+ * ```
+ * // Access a builder preconfigured with rules already existing on a given
+ * // bucket.
+ * use Google\Cloud\Storage\StorageClient;
+ *
+ * $storage = new StorageClient();
+ * $bucket = $storage->bucket('my-bucket');
+ * $lifecycle = $bucket->currentLifecycle();
+ * ```
+ *
+ * ```
+ * // Or get a fresh builder by using the static factory method.
+ * use Google\Cloud\Storage\Bucket;
+ *
+ * $lifecycle = Bucket::lifecycle();
+ * ```
+ *
+ * @see https://cloud.google.com/storage/docs/lifecycle Object Lifecycle Management API Documentation
+ */
+class Lifecycle
```

Comment:
If Lifecycle extended ArrayObject instead it could keep a similar behavior to what `$bucket->lifecycle()` provides now with additional support for AddDeleteRule and AddSetStorage helpers. 

What input @dwsupplee do you have on this design?

Reply:
I'm open to that :), please see https://github.com/googleapis/google-cloud-php/pull/1318/commits/7aeb2f832ebdffdbbfc49e6d15be56e9bf637ca5 (implemented `ArrayAccess` and `IteratorAggregate`).

Reply:
A quick sample of what is possible with the above commit:

```php
$lifecycle = Bucket::lifecycle();

// exposed via `ArrayAccess`
$lifecycle[] = [
    'action' => [
        'type' => 'Delete'
    ],
    'condition' => [
        'age' => 100
    ]
];

$lifecycle->addDeleteRule([
    'age' => 25,
    'isLive' => true
]);

// exposed via `IteratorAggregate`
foreach ($lifecycle as $rule) {
    print_r($rule);
}

Reply:
Thanks @dwsupplee!

Does this seem reasonable to you from your dev expectations? @tmatsuo do you have thoughts on this as well?

I like this design to keep Array conventions with additional helpers so developers can manipulate the lifecycle within convention and use helpers when needed.

Reply:
Yeah, I think it provides some nice value.

I especially like how it allows users to remove rules by index quite easily as opposed to only going through `clearRules`.

```php
$lifecycle = Bucket::lifecycle();
$lifecycle->addDeleteRule([
    'age' => 25,
    'isLive' => true
]);
unset($lifecycle[0]); // The lifecycle now is empty
```